src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+        } else {
+            None
+        }
+    }
+
+    fn handle_ivm(&mut self, ivm: Ivm) -> Result<(), TextFormatError> {
+        if ivm.0 == 1 && ivm.1 == 0 {
+            self.current_table.replace(CurrentSymbolTable::SystemV1);
+            Ok(())
+        } else {
+            Err(TextFormatError::UnsupportedVersion(ivm.0, ivm.1))
+        }
+    }
+
+    fn handle_meta_values(
+        &mut self,
+        value: ion::Value,
+        next: Option<ion::Value>,
+    ) -> Option<<Self as Iterator>::Item> {
+        let value = if is_system_value(&value) {
+            None
+        } else if let Some(table) = self.as_local_symbol_table(&value) {
+            match update_current_symbol_table(
+                &mut self.current_table.borrow_mut(),
+                &Some(table.clone()),
+            ) {
+                Ok(_) => None,
+                Err(e) => Some(Err(Err::Failure(IonError::from_symbol_error(
+                    self.remaining,
+                    e,
+                )))),
+            }
+        } else if let Some(_table) = self.as_shared_symbol_table(&value) {
+            // TODO: handle shared symbol tables
+            None
+        } else {
+            self.next = next;
+            Some(Ok((self.remaining, value)))
+        };
+
+        match value {
+            Some(v) => Some(v),
+            None => self.next(),
+        }
+    }
+}
+
+/// ValueIterator contains the logic for top_level within the next method.
+///
+/// note that EOF is a concept for the grammar, technically Ion streams
+/// are infinite
+/// top_level
+///     : (ws* top_level_value)* ws* value? EOF
+///     ;
+impl<'a> Iterator for ValueIterator<'a> {
+    type Item = IonResult<&'a str, ion::Value>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(next) = self.next.take() {
+            return Some(Ok((self.remaining, next)));
+        }
+
+        if self.remaining.is_empty() {
+            return None;
+        }
+
+        let maybe_ivm = take_ivm(self.remaining);
+        match maybe_ivm {
+            Ok((remaining, ivm)) => {
+                self.remaining = remaining;
+                return match self.handle_ivm(ivm) {
+                    Ok(_) => self.next(),
+                    Err(e) => Some(Err(Err::Failure(IonError::from_format_error(
+                        self.remaining,
+                        FormatError::Text(e),
+                    )))),
+                };
+            }
+            Err(Err::Failure(e)) => return Some(Err(Err::Failure(e))),
+            _ => (),
+        }
+
+        let maybe_tlv =
+            preceded(eat_ws, take_top_level_value(self.current_table.clone()))(self.remaining);
+        match maybe_tlv {
+            Ok((remaining, (value, next))) => {
+                self.remaining = remaining;
+                return self.handle_meta_values(value, next);
+            }
+            Err(Err::Failure(e)) => return Some(Err(Err::Failure(e))),
+            _ => (),
```

Comment:
It is intentional actually, and stems from Nom's behavior.

When processing the token stream I'm checking for IVMs here since there are many rules about the exact representation of a valid IVM. Most of that information goes away when we convert it into a Symbol, so we would need a special Symbol type for IVMs if we wanted to capture them when parsing the actual identifier symbol.

An `Err(Err::Failure(..))` indicates an unrecoverable failure whereas `Err(Err::Error(..))` just indicates that Nom couldn't apply the provided parser. Since in this case we just want to fall through if the item isn't an IVM, we ignore the error and move on. Note that we include the error from the final parser combinator in the result, which is the most lax of the parsers we have attempted to apply.

I can add comments to maybe_ivm and maybe_tlv to clarify this if that would help.

This does make me realize that it is possible for maybe_last to apply multiple times over the course of the iterator which is counter to the spec (it should be applied at most once, to the very last item, which doesn't have the same delimiting constraints that other items have). I'll think about how to fix this.

Reply:
Your explanation makes sense. Nom's parsers can produce odd looking code when used outside of combinators. A comment would be helpful, since I can't think of a better way to structure this other than trying to find a way to fit all of the logic from here to the end of the function into an `alt` combinator.

Reply:
> ... trying to find a way to fit all of the logic from here to the end of the function into an alt combinator

Brilliant! This might actually be nicer.

Reply:
After staring at it for a while, I'm thinking it is simpler the way it is unless we decide to not use an iterator here. I've been waffling back and forth on that, what are your thoughts?

Reply:
I think it's fine the way it is. Shifting away from the Iterator would be a significant refactoring, and I don't think a decision needs to be made now.